src/v/datalake/record_schema_resolver.cc

```diff
@@ -0,0 +1,206 @@
+        const auto& avro_schema = avro_def();
+        try {
+            // TODO: we should use Avro value conversion instead of this
+            // Iceberg-specific code.
+            auto type = iceberg::type_from_avro(
+              avro_schema.root(), iceberg::with_field_ids::no);
+            co_return type_and_buf{
+              .schema = avro_schema,
+              .pb_offsets = std::nullopt,
+              .type = std::move(type),
+              .parsable_buf = std::move(buf_no_id),
+            };
+        } catch (...) {
+            vlog(
+              datalake_log.error,
+              "Avro schema translation failed: {}",
+              std::current_exception());
+            co_return errc::schema_translation_error;
+        }
+    }
+    ss::future<checked<type_and_buf, errc>>
+    operator()(const ppsr::protobuf_schema_definition&) {
+        // TODO: will be implemented in a subsequent commit!
+        co_return errc::not_supported;
+    }
+    ss::future<checked<type_and_buf, errc>>
+    operator()(const ppsr::json_schema_definition&) {
+        co_return errc::not_supported;
+    }
+};
+
+} // namespace
+
+resolved_buf resolved_buf::make_raw_binary(iobuf b) {
+    return resolved_buf{
+      .schema_identifier = std::nullopt,
+      .type = std::nullopt,
+      .val = binary_val(std::move(b)),
+    };
+}
+
+ss::future<checked<resolved_buf, record_schema_resolver::errc>>
+record_schema_resolver::resolve_buf_schema(iobuf b) const {
+    schema_message_data schema_id_res;
+    try {
+        // NOTE: Kafka's serialization protocol relies on a magic byte to
+        // indicate if we have a schema. This has room for false positives, and
+        // we can't say for sure if an error is the result of the record not
+        // having a schema. Just translate to binary.
+        auto res = get_value_schema_id(b);
+        if (res.has_error()) {
+            co_return resolved_buf::make_raw_binary(std::move(b));
+        }
+        schema_id_res = std::move(res.value());
+    } catch (...) {
+        co_return resolved_buf::make_raw_binary(std::move(b));
+    }
+    auto schema_id = schema_id_res.schema_id;
+    auto buf_no_id = std::move(schema_id_res.shared_message_data);
+
+    // TODO: It'd be nice to cache these -- translation interval instills a
+    // natural limit to concurrency so a cache wouldn't grow huge.
+    auto schema_fut = co_await ss::coroutine::as_future(
+      sr_.get_valid_schema(schema_id));
+    if (schema_fut.failed()) {
+        vlog(
+          datalake_log.warn,
+          "Error getting schema from registry: {}",
+          schema_fut.get_exception());
+        // TODO: make schema::registry tell us whether the issue was transient.
+        co_return errc::registry_error;
+    }
+    auto resolved_schema = std::move(schema_fut.get());
+    auto typed_buf_res = co_await resolved_schema.visit(
+      schema_translating_visitor{std::move(buf_no_id)});
+    if (typed_buf_res.has_error()) {
```

Comment:
maybe worth a log message

Reply:
Shuffled this code around, there should be an error logged now

Reply:
Done